src/gap_to_julia.jl

```diff
@@ -78,164 +94,153 @@ function gap_to_julia(::Type{Rational{T}}, x::Int64) where T <: Integer
+## Vectors
+function gap_to_julia(::Type{Vector{T}}, obj::GapObj, recursion_dict=IdDict(); recursive=true) where T
+    ! Globals.IsList(obj) && throw(ConversionError(obj, Vector{T}))
+    if ! haskey(recursion_dict, obj)
+        len_list = length(obj)
+        new_array = Vector{T}(undef, len_list)
+        recursion_dict[obj] = new_array
+        for i in 1:len_list
+            current_obj = ElmList(obj, i)  # returns 'nothing' for holes in the list
+            if recursive
+                new_array[i] = get!(recursion_dict, current_obj) do
+                    gap_to_julia(T, current_obj, recursion_dict;
+                                 recursive=true)
+                end
+            else
+                new_array[i] = current_obj
+            end
         end
     end
-    return new_array
+    return recursion_dict[obj]
 end
 
-## Matrix / list-of-lists
-function gap_to_julia( ::Type{Array{T,2}}, obj :: GapObj, recursion_dict = IdDict() ) where T
-    if ! Globals.IsList( obj )
-        throw(ArgumentError("<obj> is not a list"))
-    end
-    if haskey(recursion_dict, obj)
+## Matrices or lists of lists
+function gap_to_julia(::Type{Matrix{T}}, obj::GapObj, recursion_dict=IdDict(); recursive=true) where T
+    if Globals.IsMatrixObj(obj)
+        if ! haskey(recursion_dict, obj)
+            elm = Globals.MatElm
+            nrows = Globals.NumberRows(obj)
+            ncols = Globals.NumberColumns(obj)
+            new_array = Matrix{T}(undef, nrows, ncols)
+            recursion_dict[obj] = new_array
+            for i in 1:nrows
+                for j in 1:ncols
+                    current_obj = elm(obj, i, j)
+                    if recursive
+                        new_array[i, j] = get!(recursion_dict, current_obj) do
+                            gap_to_julia(T, current_obj, recursion_dict;
+                                         recursive=true)
+                        end
+                    else
+                        new_array[i, j] = current_obj
+                    end
+                end
+            end
+        end
         return recursion_dict[obj]
-    end
-    len_list_outer = length(obj)
-    len_list_inner = len_list_outer == 0 ? 0 : length(obj[1])
-    new_array = Array{T,2}( undef, len_list_outer, len_list_inner )
-    recursion_dict[obj] = new_array
-    for i in 1:len_list_outer
-        for j in 1:len_list_inner
-            current_obj = ElmList(ElmList(obj, i), j)
-            new_array[i, j] = get!(recursion_dict, current_obj) do
-                gap_to_julia(T,current_obj, recursion_dict)
+    elseif Globals.IsList(obj)
+        if ! haskey(recursion_dict, obj)
+            nrows = length(obj)
+            ncols = nrows == 0 ? 0 : length(obj[1])
+            new_array = Matrix{T}(undef, nrows, ncols)
+            recursion_dict[obj] = new_array
+            for i in 1:nrows
+                for j in 1:ncols
+                    current_obj = ElmList(ElmList(obj, i), j)
```

Comment:
well, this does return `nothing`

Reply:
The term GAP matrix is perhaps misleading.
The conversion in question shall deal with GAP objects such that the result is a Julia matrix (2-dim. array).
The candidates on the GAP side are `IsMatrixObj` objects, which are in general not in `IsList` and which do in general not support nested `ElmList` access, and `IsList` objects whose entries are again `IsList` objects, which are in general not `IsMatrixObj` objects.
The former ones cannot have holes, the latter ones can.

Depending on the prescribed Julia type, it may work to convert a nested GAP list with holes to a Julia matrix that contains `nothing` in some places; one of the tests looks as follows.
```
n = GAP.EvalString("[[1,2],[,4]]")
@test GAP.gap_to_julia(Array{Union{Int64,Nothing},2},n) == [1 2; nothing 4]
```